[RunAllTests] Fix part of #2787: Add support for proto lesson loading from local assets

BenHenning opened the pull request:
Fix part of #2787

This adds initial support for loading the new lessons needed for alpha. In particular, to better prepare the codebase for post-alpha beta work, this PR is introducing support for directly loading proto assets rather than JSON (bypassing the entire JSON loading pipeline). This requires punching a few holes into various retrievers/controllers (gated by a new qualified boolean: LoadLessonProtosFromAssets) to load proto rather than JSON. The resulting proto code is a bit simpler than the JSON versions since typing becomes much simpler.

In order to ensure the new pathways are (mostly) working in tests, this PR enables proto caching in some tests. However doing this isn't straightforward: we don't want to check actual binary proto files into the codebase since they are binary files and will  unnecessarily bloat our repository. Textproto is a more reasonable alternative since it's plaintext (and actually a bit more compact than JSON), but complicated to use: proto lite does not support loading via TextFormat. Rather than introducing a dependency on JsonFormat (to avoid JSON altogether), I decided instead to introduce a custom Starlark rule in Bazel to, at build time, convert textproto files to binary files & include those at build time. This works quite well: for both the apk itself & tests we can load binary pb files as though we're including them directly. Also, for this reason is why the PR includes textproto conversions of our development assets.

However, the above approach only works for Bazel. To mitigate this, the few test suites that were updated to verify the fixes conditionally enable proto loading based on whether the environment is Bazel. This is a bit hacky, but it seems like the best solution in the interim until we can entirely move off of Gradle per #59.

Other considerations for this PR:
- This removes dependence on test story 1 entirely because it includes structural elements that are incompatible with the pipeline used to convert to text/binary proto. This seems fine since we don't actually need these interactions for testing or development.
- This removes support for confirmed unclassified answers since they aren't used at all, and probably won't be in Android.
- This PR does not introduce support for the up-to-date interactions that will be shipped with the alpha. That will happen in a later PR (mainly #2985).
- The textproto files introduced in this PR were generated using an in-progress script that's also being used to download the assets that will be used for alpha. Follow up with me if you want to see the script.
- Enabling proto caching & running the APK locally will yield to images not loading. This is partly due to #2979 (in that they can't load without that fix), but mostly due to the fact that all of our dev assets short-circuit thumbnails (which is why we never discovered #2940 until working on alpha)

BenHenning committed:
Add support for loading proto versions of lessons.

BenHenning committed:
Upgrade Espresso versions to work around build issues.

Note that I can't actually verify Espresso tests are working with these
versions since the Bazel project doesn't yet have Espresso tests set up.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Lint fix post-copying from #2927.

BenHenning committed:
Remove image loading annotation.

This will be added in a later branch in the chain.

BenHenning committed:
Add caching module to needed test suites.

BenHenning commented:
/cc @vinitamurthi

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/topic/TopicController.kt:
```diff
@@ -396,12 +426,25 @@ class TopicController @Inject constructor(
 
+  private fun loadSubtopic(topicId: String, subtopicId: Int): Subtopic {
+    val subtopicRecord = assetRepository.loadProtoFromLocalAssets(
+      assetName = "${topicId}_$subtopicId",
```
I'm wondering if there's a better way to sent this request rather than the assetname which is a slightly vague variable name since it looks like it's picking up assets for different ids. Maybe asset type (as an enum), asset Id (which is called asset name here) , and base message?

BenHenning replied:
Not sure I follow--what advantage does an enum give us here? The main typing mechanism is the base message type, and introducing an enum may require us to keep the types & enum in-sync which will add complexity.

vinitamurthi replied:
My suggestion was because we are setting the asset name to be of a different format based on the asset types (i.e. subtopics have a different format of asset name (being topicId_subtopicID) , where as other assets have a different format for their asset names. But it looks like this is because of the way the file names are created, and given that this code may be unnecessary in a few months I guess its okay.

BenHenning replied:
Yeah--I don't think we should over-index on this since the long-term solution should utilize more of a database-type structure for managing files. Thanks for bringing this up, though!

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:
```diff
@@ -123,6 +130,23 @@ class TopicListController @Inject constructor(
+    return if (loadLessonProtosFromAssets) {
+      val topicIdList =
+        assetRepository.loadProtoFromLocalAssets(
+          assetName = "topics",
```
Here this kind of breaks the paradigm for assetName compared to the fields that are being used above, so I think it may be better to keep assetType, assetId, and baseMessage

BenHenning replied:
It's not a type though, it's the actual file name checked into the APK. Calling it 'assetType' would be misleading. Perhaps 'assetBaseFIilename' would be the most clear, but it's a bit lengthy. I also don't understand how you're distinguishing between 'type' and 'ID'. This is the topic list so it doesn't have an ID.

It could be named better but I'm following existing filenames to keep things a bit simpler. We'll be removing all of this code in coming months, anyway, in favor of downloading & caching completely dynamically.

vinitamurthi replied:
Same comment as above..given that this code will be dead soon I think its okay

BenHenning replied:
Ack, thanks.

BenHenning committed:
Add support for importing text protos as assets.

This actually provides Bazel-only support for converting text protos to
binary and including those as assets so that we can avoid checking in
binary protos to the codebase.

BenHenning committed:
Add gitignore for Android Studio Bazel plugin output.

BenHenning committed:
Import textproto versions of existing dev assets.

This imports local conversions of all existing dev assets as text proto
leveraging the Bazel-only conversion system to convert these to binary
protos and include them as assets.

This is a necessary step to both eventually replace JSON assets with
text proto, and to test that the new proto loading pipeline is working
as designed.

BenHenning committed:
Undo Espresso version change since it breaks Bazel tests.

BenHenning committed:
Fix tests.

Various attempts to fix existing tests + move some of the domain tests
over to actually using proto when running with Bazel.

App module tests are hanging currently for unknown reasons (only in
Bazel).

BenHenning committed:
Use correct Espresso core.

3.1.0 causes tests to hang indefinitely when run with Bazel.

BenHenning committed:
Move coroutine dispatchers.

This moves the coroutine dispatchers to a new threading subpackage. It
also simplifies the TestCoroutineDispatcher interface & implementations
(which required moving runUntilIdle to CoroutineExecutorServiceTest).

The test coroutine annotations were also moved to their own files.

BenHenning committed:
Refactor testing utilities build graph.

This introduces new Bazel libraries for:
- Test dispatchers
- Fake system/Oppia clocks
- Robolectric dependencies

BenHenning committed:
Introduce dedicated test for TestCoroutineDispatcher.

This also fixes some threading issues in the dispatcher, and clarifies
some of its API that was previously unclear (and can lead to subtle race
conditions).

BenHenning committed:
Add tests for TestCoroutineDispatcher.

This required a bunch of refactoring, adding the first examples of
dedicated Bazel test build files, and introducing a new test pattern for
sharing code between different implementations. Further, it refined the
API for TesCoroutineDispatcher & fixed some issues in the
implementations (especially Robolectric).

Tests verified as non-flaky in Robolectric across 1000 runs, and 0.2%
flaky across 1000 runs for Espresso (though that may be fixed; waiting
on a follow-up run to confirm stability).

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Workaround ktlint semicolon issue.

BenHenning committed:
Lint fixes.

BenHenning committed:
Fix test post-merge.

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Move CoroutineDispatcher to be part of threading.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into fix-protobuf-duplication-issue

BenHenning committed:
Undo version upgrade.

BenHenning committed:
Merge branch 'fix-protobuf-duplication-issue' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	.gitignore
	app/src/sharedTest/java/org/oppia/android/app/story/StoryActivityTest.kt

BenHenning committed:
Fix broken tests.

BenHenning committed:
Lint fixes.

BenHenning committed:
Add a bit more proto loading test coverage.

BenHenning committed:
Comment fix.

BenHenning commented:
NB: I will change this & downstream PRs to use '[RunAllTests]' to force test runs since these PRs appear to have issues triggering the 'compute affected tests' flow (probably because of changing BUILD files--that sometimes causes the script to hang and I haven't yet figured out why).

BenHenning committed:
Lint fixes.

BenHenning committed:
Address earlier TODO.

BenHenning commented:
Thanks @vinitamurthi. FYI that this is ready for full review--PTAL.

@anandwana001 & @rt4914 PTAL as codeowners.

rt4914 reviewed (approved):
LGTM for code-owner files.

BenHenning committed:
Merge branch 'develop' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Fix broken test build.

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/story/StoryFragmentTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
	testing/src/test/java/org/oppia/android/testing/story/StoryProgressTestHelperTest.kt

anandwana001 commented on domain/domain_assets.bzl:
```diff
@@ -0,0 +1,213 @@
+
+def _extract_proto_sources(deps):
+    """
+    Returns the list of proto source files that make up the specified list of proto dependencies.
```
we can use the `#` for comments, nit fix, not blocking

BenHenning replied:
@anandwana001 I don't think we should be. See https://skydoc.bazel.build/docs/writing.html for examples on writing Skydocs. They use multi-line strings, not ``#``. I think we should be treating ``#`` as line comments and ``"""`` as documentation (which also matches Oppia web).

I did notice I'm not formatting on the correct line, but we can fix that in a later PR by making all documentation consistent.

anandwana001 reviewed (approved):
LGTM, nit change but can be ignored

BenHenning committed:
Merge branch 'develop' into add-coroutine-dispatcher-tests

BenHenning committed:
Merge branch 'add-coroutine-dispatcher-tests' into add-support-for-proto-lesson-loading-from-local-assets

BenHenning commented:
ClassroomService Bazel failure will be fixed by #2916.

BenHenning commented:
Thanks all for the review! Merging this now to keep the PR chain moving forward.
